provider/vsphere/ova_utils.go

```diff
@@ -5,79 +5,165 @@ package vsphere
 
 import (
 	"archive/tar"
+	"crypto/sha256"
+	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"os"
 	"path/filepath"
 
 	"github.com/juju/errors"
+	"github.com/juju/mutex"
 )
 
-func downloadOva(basePath, url string) (string, error) {
-	logger.Debugf("Downloading ova file from url: %s", url)
-	resp, err := http.Get(url)
+// CacheLocker is an interface that may be used for locking
+// an OVA cache directory.
+type CacheLocker interface {
+	// Lock locks the OVA cache directory for concurrent access,
+	// and returns a function to release the lock or an error.
+	Lock() (func(), error)
+}
+
+type mutexCacheLocker struct {
+	spec mutex.Spec
+}
+
+// NewMutexCacheLocker returns an implemenation of CacheLocker
+// based on juju/mutex, using the given mutex.Spec.
+func NewMutexCacheLocker(spec mutex.Spec) CacheLocker {
+	return mutexCacheLocker{spec: spec}
+}
+
+// Lock is part of the CacheLocker interface.
+func (l mutexCacheLocker) Lock() (func(), error) {
+	releaser, err := mutex.Acquire(l.spec)
+	if err != nil {
+		return nil, err
+	}
+	return releaser.Release, nil
+}
+
+// downloadOVA downloads and extracts the OVA identified by the
+// given image metadata into the specified directory. We store
```

Comment:
Is using a directory the best choice? Previously when we supported caching LXC images, we used the mongo blob store because in a HA scenario, any controller could service the request. Using a directory means each controller will have it's own cache. Why not cache the images in mongo?